### ISSUE TITLE
Education state transitions

### ISSUE THREAD

**Issue by ld-archer** [Leeds-MRG/Minos#271]
Currently, a respondents max education state is predicted before the pipeline begins, during production of the stock and replenishing populations (after estimating the education state model). Individuals are then deterministically transitioned to higher education states at specific ages. This is a big assumption in and of itself, but more work to get this closer to reality would most likely have a negligible effect on our population and even smaller on the outcomes. 

The issue with this is to handle the move from deterministic transitions to the model based predictions in the S7_labour_state variable. The labour state variable has a factor level for full time education, which needs to work in partnership with the education_state variable and the education module. What we can do is force the labour state variable to FT education until a respondent reaches their max education level, and then move to transitioning their labour state based on the transition model. This will probably mean some people are transitioned back into FT education throughout the model, but hopefully this number will be small enough to be negligible (fewer numbers still in education as age increases in raw data). 

Tasks:
- [x] Check what age people achieve each level of education
    - [x] Average age plot (see if change over time in raw)
    - [x] Histograms
    - [ ] Animated histograms?
- [x] Force labour_state to FT education until max_educ has been reached (this might already have been done)
- [x] Change labour_state prediction to only run for those who have reached max_educ
- [x] Check how many people move into FT education after age of 30
- [x] Make sure stock population is doing max_educ prediction
- [ ] IF all the above goes well, add S7_labour_state variable into the default income -> SF12 experiment, so we can use the same education functionality in there also. Plus labour_state is an important variable there that is currently missing.

**Comment by ld-archer** [comment]
## Education Level Age Check

Sticking some plots in here just for reference.

### Age Achieve Education Levels
![education_state_1](https://github.com/Leeds-MRG/Minos/assets/44437206/9b561ec7-bed9-48fd-8c81-a1d6e724eed5)
![education_state_2](https://github.com/Leeds-MRG/Minos/assets/44437206/19590a38-8129-4f0a-8fdc-99f78c1ef77d)
![education_state_3](https://github.com/Leeds-MRG/Minos/assets/44437206/62014c3d-eea7-4828-b7dc-8589d2e95c3f)
![education_state_5](https://github.com/Leeds-MRG/Minos/assets/44437206/f79c1590-e260-4e97-8f07-4cc9f888782e)
![education_state_6](https://github.com/Leeds-MRG/Minos/assets/44437206/5c7b06b3-adda-4bf3-9636-b66caef079e1)
![education_state_7](https://github.com/Leeds-MRG/Minos/assets/44437206/c1a6c37d-b42e-4947-96a3-10cb695b51d4)

#### Mean Age Achieved

![mean_age_achieved_educ](https://github.com/Leeds-MRG/Minos/assets/44437206/112e89aa-7664-4f90-8e35-fd07f3f0e109)

**Comment by ld-archer** [comment]
Note I wrote in the education module about when education transition happen. The TL;DR for this is that some decisions are made based on knowledge of the education system (GCSEs and A-levels are somewhat predictable), and others are made by using the average age they are achieved in Understanding Society data. There is also a bit of a trade off with level 7 qualifications (masters/PhD and equivalent). The average age these are achieved in data is 32, but this would mean individuals are in FT education from 16-32. This is a bit much, as they are probably in employment for a chunk of that time so I've pulled this back to age 30.

#### NOTE ####
        # The age that these qualifications are achieved varies quite a lot in Understanding Society data, but trying
        # to include that variance here would add a level of complexity that is probably not worth the hassle (we're
        # assuming that the long term effects of education on health are more important than if an individual achieved
        # e.g. a level 7 qualification later in life).
        # Therefore, we are making a couple of assumptions on when to transition based on either knowledge of the
        # UK education system, or the average age at which these levels are achieved in the underlying data. Note that
        # the age achieved is 1 year after the qualifications are held to account for the range of dates that
        # individuals answer the survey. These are:
        # Level 1 and 2 - Age 17
        #   1 & 2 are equivalent to GCSEs at different grades, and because young people are required to be in full-time
        #   education until 16 it seems reasonable to assume that everyone will achieve this level.
        # Level 3 - Age 19
        #   Level 3 is equivalent to A-levels, advanced apprenticeships, and baccalaureate's. Government guidelines do
        #   require young people to be in FT education, apprenticeship, or combination of volunteering and PT education,
        #   but there is no guarantee that they will achieve these levels so this is not guaranteed like level 1 or 2.
        # Level 5 - Age 30
        #   Nursing or med quals are a weird one, so this is based on the average age achieve in US as there is large
        #   variance in the data.
        # Level 6 - Age 27
        #   Again basing this on US data as it covers first class degrees but also a range of other qualifications.
        # Level 7 - Age 30
        #   This is an unfortunately difficult decision, and is a bit of a balance between the mean age in US data and 
        #   a 'want' to not have respondents in full time education for too long in the model. The mean age this level 
        #   is achieved is 32 in the data, but that would mean these people are in FT education from 16-32 which is 
        #   not realistic. I think these individuals would most likely have some time in employment between these ages
        #   but that would be very complex to implement.

**Comment by ld-archer** [comment]
Struggling with something now that might require a bit of a change in approach.

A problem from the start has been that the proportion of the sample in full time education creeps up throughout the model, replacing those in the Not Working category (long-term disabled, retired, probably other categories). This plot illustrates this:
![non-deterministic_retirement](https://github.com/Leeds-MRG/Minos/assets/44437206/6d21e625-2823-4114-902d-e4f31e242747)

Initially I wondered if this was due to retirement not being a well-predicted thing, so I made retirement a deterministic thing also with 65 being the retirement age (maybe I should make it 67 now?). After changing this setting so that all 65 and older moved into the Not Working category, this was the result:
![deterministic_retirement](https://github.com/Leeds-MRG/Minos/assets/44437206/ace661f7-5300-44bc-b72b-49a01380dee7)

This has resulted in the Not Working group remaining fairly steady, so not handling retirement was the issue for this group, but now the FT Education group is still trending up in the same way but to the detriment of the FT Employed. I don't have a fully baked solution in mind for this yet, but am going to try a few things and see what happens. One thing I can cross off the list is changing the ages at which people reach certain levels of education. I initially reduced the age to achieve levels 5-7 hoping that would change the outcomes, but it had very little impact on these plots.

The only thing I can think of doing right now that doesn't require scrapping what we have and recreating is to allow people to join the workforce between reaching education levels, and maybe increase the ages of level 5 and 7. Level 5 is nursing / medical qualifications and the mean age of achieving these qualifications is around 31, perhaps we could assume these people do not stay in full time education for that length, but instead get these qualifications whilst in the workforce.
